cdap-ui/app/directives/timeline/timeline.js

```diff
@@ -115,78 +131,226 @@ function link (scope, element) {
+        .innerTickSize(-40)
+        .tickPadding(7)
+        .ticks(8);
 
     //Generate circles from the filtered events
     let circles = svg.selectAll('circle')
-      .data(filteredEvents)
+      .data(timelineData)
       .enter()
       .append('circle');
 
-    let xScale = d3.time.scale()
-                         .domain([minDate, maxDate])
-                         .nice(d3.time.minute)
-                         .range([0,width]);
-
-    let xAxis = d3.svg.axis()
-      .orient('bottom')
-      .scale(xScale);
-
-    //Initially renders circles uniformally across timeline
-    circles.attr('cx', function(d, i) {
-              return i*50 + 25;
-            })
-            .attr('cy', height-height/3)
-            .attr('r', 3)
-            .attr('class', function(d) {
-              if(d.level === 'ERROR'){
-                return 'red-circle';
-              }
-              else if(d.level === 'WARN'){
-                return 'yellow-circle';
-              }
-            });
+    circles.attr('cx', function(d) {
+      let xVal = Math.floor(xScale(d.time));
+      if(timelineStack[xVal] === undefined){
+        timelineStack[xVal] = 0;
+      } else {
+        timelineStack[xVal]++;
+      }
+      return xScale(d.time) + 15;
+    })
+    .attr('cy', function(d) {
+      let numDots = timelineStack[Math.floor(xScale(d.time))]--;
+      return height-height/2.5 - (numDots * 6);
+    })
+    .attr('r', 2)
+    .attr('class', function(d) {
+      if(d.level === 'ERROR'){
+        return 'red-circle';
+      }
+      else if(d.level === 'WARN'){
+        return 'yellow-circle';
+      } else {
+        return 'other-circle';
+      }
+    });
 
+    // -------------------------Build Brush / Sliders------------------------- //
+    //X-Axis
     svg.append('g')
-      .attr('class', 'xaxis')
-      .attr('transform', 'translate(0,' + (height - 10) + ')')
+      .attr('class', 'xaxis-bottom')
+      .attr('transform', 'translate(' + ( (paddingLeft + paddingRight) / 2) + ',' + (height - 20) + ')')
       .call(xAxis);
-  }
+
+    //attach handler to brush
+    brush = d3.svg.brush()
+        .x(xScale)
+        .extent([0,0])
+        .on('brush', function(){
+          if(d3.event.sourceEvent) {
+            let v = xScale.invert(d3.mouse(this)[0]);
+            let index = d3.mouse(this)[0];
+            if(v !== leftVal){
+              leftVal = v;
+            }
+            updateSlider(index);
+          }
+        });
+
+    //Fix me: Make me self reliant (see 'xScale' and other variables)
+    function updateSlider(val) {
+      //Update the brush position
+      if(val < 0){
+        val = 0;
+      }
+      if(val > sliderLimit){
+        val = sliderLimit;
+      }
+
+      sliderX = val;
+
+      if(sliderX > pinX){
+        updatePin(sliderX);
+      }
+
+      leftHandle.attr('x', val);
+      console.log('Left slider is at : ' + xScale.invert(val));
+
+      sliderBar.attr('d', 'M0,0V0H' + val + 'V0');
```

Comment:
This should eventually call `updateTheSlider` or appropriate function that updates the `LogViewerStore`.